thrift/thrift-gen/tchannel-template.go

```diff
@@ -142,46 +149,59 @@ func (s *{{ .ServerStruct }}) Handle(ctx {{ contextType }}, methodName string, p
 }
 
 {{ range .Methods }}
-	func (s *{{ $svc.ServerStruct }}) {{ .HandleFunc }}(ctx {{ contextType }}, protocol athrift.TProtocol) (bool, athrift.TStruct, error) {
+	func (s *{{ $svc.ServerStruct }}) {{ .HandleFunc }}(ctx {{ contextType }}, protocol athrift.TProtocol) (handled bool, resp athrift.TStruct, retErr error) {
 		var req {{ .ArgsType }}
 		var res {{ .ResultType }}
+		const serviceMethod = "{{ $svc.ThriftName }}::{{ .ThriftName }}"
 
-		if err := req.Read(protocol); err != nil {
-			return false, nil, err
+		if readErr := req.Read(protocol); readErr != nil {
+			return false, nil, readErr
 		}
 
-		{{ if .HasReturn }}
-			r, err :=
-		{{ else }}
-			err :=
-		{{ end }}
-				s.handler.{{ .Name }}({{ .CallList "req" }})
-
-		if err != nil {
-			{{ if .HasExceptions }}
-			switch v := err.(type) {
-				{{ range .Exceptions }}
+		postRun, err := s.interceptorRunner.RunPre(ctx, serviceMethod, &req)
+
+		defer func () {
+			resp = &res
+			retErr = postRun(resp, err)
+			handled = retErr == nil
+			if retErr != nil {
+				resp = nil
+				{{ if .HasExceptions }}
+				switch v := retErr.(type) {
+					{{ range .Exceptions }}
 					case {{ .ArgType }}:
 						if v == nil {
-							return false, nil, fmt.Errorf("Handler for {{ .Name }} returned non-nil error type {{ .ArgType }} but nil value")
+							retErr = fmt.Errorf("Handler for {{ .Name }} returned non-nil error type {{ .ArgType }} but nil value")
```

Comment:
nit: not changed by you, but errors are supposed to be lowercase, can you make `Handler` `handler`